cdi/src/main/java/org/hibernate/validator/cdi/internal/ValidatorBean.java

```diff
@@ -44,6 +46,11 @@ public ValidatorBean(BeanManager beanManager, Bean<?> validatorFactoryBean,
 		this.types = Collections.unmodifiableSet(
 				CollectionHelper.<Type>newHashSet(
 						ClassHierarchyHelper.getHierarchy( validationProviderHelper.getValidatorBeanClass() )
+								.stream()
+								// We do not include impl class of a validator if it comes from Hibernate Validator.
+								// In this case we only want to share the interfaces:
+								.filter( c -> !( validationProviderHelper.isHibernateValidator() && c.equals( validationProviderHelper.getValidatorBeanClass() ) ) )
```

Comment:
This seems... fragile?

If we add some abstract class above `ValidatorImpl`, it will probably no longer work.

And with `ValidatorProviderHelper.forDefaultProvider`, it will definitely not work -- though I admit I have no idea what's the purpose of that thing.

Shouldn't you adjust how `ValidationProviderHelper` is built instead? E.g. `forHibernateValidator` would pass `HibernateValidatorFactory.class`/`HibernateValidator.class` to the constructor, while `forDefaultProvider` would do that if `ValidatorFactory` implements `HibernateValidatorFactory`, and default to some other behavior if not (maybe `ValidatorFactory.class`/`Validator.class`?).

Reply:
hmm, if we pass just the interface, then we might lose some interfaces:
```java
ValidatorImpl implements Validator, ExecutableValidator
```
On the other hand, the `ValidatorFactoryImpl` extends `AutoCloseable`, and since we are taking the class hierarchy, we add `AutoCloseable` to the types, so that means we allow something like ...
```
@Inject
@HibernateValidator
AutoClosable validatorFactory;
```

is possible? 🤔 😃 

Maybe... if it is an HV factory/validator, we filter out any internal classes/interfaces by looking at the package, and otherwise, we keep everything. There's this test that checks if someone can inject their own custom factory/validator, and in that case, the implementation classes are injected, that's why I think if it is not "ours", we just keep it all and leave it to the user to deal with.

https://github.com/hibernate/hibernate-validator/blob/e5a569e122330feb37677a5a33d6322948a819b1/cdi/src/test/java/org/hibernate/validator/test/cdi/internal/injection/InjectionOfCustomProviderTest.java#L51-L73

Reply:
> if it is an HV factory/validator, we filter out any internal classes/interfaces by looking at the package

Okay, but perhaps hardcode that list (with a test to check it stays up to date).

> otherwise, we keep everything. There's this test that checks if someone can inject their own custom factory/validator, and in that case, the implementation classes are injected, that's why I think if it is not "ours", we just keep it all and leave it to the user to deal with.

Makes sense.